Box the biggest ast::ItemKind variants

bugadani opened the pull request:
This PR is a different approach on https://github.com/rust-lang/rust/pull/81400, aiming to save memory in humongous ASTs.

The three affected item kind enums are:
 - `ast::ItemKind` (208 -> 112 bytes)
 - `ast::AssocItemKind` (176 -> 72 bytes)
 - `ast::ForeignItemKind` (176 -> 72 bytes)

Mark-Simulacrum commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

bors commented:
:hourglass: Trying commit 7b379a25a2ac7d8dd3508f73e5cdec3e8a34452d with merge 7fc85fb97c995b5a91e77a8c4addcd865da24d11...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7b379a25a2ac7d8dd3508f73e5cdec3e8a34452d","merge_sha":"7fc85fb97c995b5a91e77a8c4addcd865da24d11"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1770232189)
Build commit: 7fc85fb97c995b5a91e77a8c4addcd865da24d11 (`7fc85fb97c995b5a91e77a8c4addcd865da24d11`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1770232189"},"merge_sha":"7fc85fb97c995b5a91e77a8c4addcd865da24d11"} -->

rust-timer commented:
Queued 7fc85fb97c995b5a91e77a8c4addcd865da24d11 with parent 1483e67addd37d9bd20ba3b4613b678ee9ad4d68, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1483e67addd37d9bd20ba3b4613b678ee9ad4d68&end=7fc85fb97c995b5a91e77a8c4addcd865da24d11).

@rustbot label: +S-waiting-on-perf

rust-timer commented:
Finished benchmarking try commit (7fc85fb97c995b5a91e77a8c4addcd865da24d11): [comparison url](https://perf.rust-lang.org/compare.html?start=1483e67addd37d9bd20ba3b4613b678ee9ad4d68&end=7fc85fb97c995b5a91e77a8c4addcd865da24d11).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

bugadani commented:
Wait, what did I do to make typeck faster? I thought that works on the HIR and not the AST o.O Anyway, small runtime improvement and a disappointing lack of improvement in max-rss.

Mark-Simulacrum commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

bors commented:
:hourglass: Trying commit ff464fe5d23cee78979bf8669c8c83ef7ffcb2b7 with merge bf1b6dbd7f67fc51be6c4ba42a6d0d25a6fa9a3b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"ff464fe5d23cee78979bf8669c8c83ef7ffcb2b7","merge_sha":"bf1b6dbd7f67fc51be6c4ba42a6d0d25a6fa9a3b"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1777625571)
Build commit: bf1b6dbd7f67fc51be6c4ba42a6d0d25a6fa9a3b (`bf1b6dbd7f67fc51be6c4ba42a6d0d25a6fa9a3b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1777625571"},"merge_sha":"bf1b6dbd7f67fc51be6c4ba42a6d0d25a6fa9a3b"} -->

rust-timer commented:
Queued bf1b6dbd7f67fc51be6c4ba42a6d0d25a6fa9a3b with parent 613ef740f3f37702728c6324f948d0abd1e9c82b, future [comparison URL](https://perf.rust-lang.org/compare.html?start=613ef740f3f37702728c6324f948d0abd1e9c82b&end=bf1b6dbd7f67fc51be6c4ba42a6d0d25a6fa9a3b).

@rustbot label: +S-waiting-on-perf

rust-timer commented:
Finished benchmarking try commit (bf1b6dbd7f67fc51be6c4ba42a6d0d25a6fa9a3b): [comparison url](https://perf.rust-lang.org/compare.html?start=613ef740f3f37702728c6324f948d0abd1e9c82b&end=bf1b6dbd7f67fc51be6c4ba42a6d0d25a6fa9a3b).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

bugadani commented:
Perf is largely the same as before, max-rss *maybe* a bit better (uh, less terrible?) if you squint real hard, but that can always be noise.

bugadani commented:
I undid most of the clippy formatting changes (thanks, RA), and I declare this ready for review.

bors commented:
:umbrella: The latest upstream changes (presumably #81560) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

cjgillot commented:
Can you add  a `static_assert_size!` on `Item`, `TraitItem`, `ImplItem` and `ForeignItem`? This will allow to avoid detect regressions.
Do you have the sizes of those structs before this PR?

cjgillot commented on compiler/rustc_ast/src/ast.rs:
```diff
@@ -2655,6 +2655,35 @@ impl Default for FnHeader {
+}
+
+#[derive(Clone, Encodable, Decodable, Debug)]
+pub struct FnKind(pub Defaultness, pub FnSig, pub Generics, pub Option<P<Block>>);
```
Should these new structs have named fields?

bugadani replied:
Honestly, I'm undecided. These were fine without before, and adding names would just bloat up a lot of match patterns. If we were to start replacing them here, we could go through the whole codebase and replace like two million tuple structs as well.

cjgillot commented on compiler/rustc_ast/src/ast.rs:
```diff
@@ -2655,6 +2655,35 @@ impl Default for FnHeader {
+    pub Generics,
+    pub GenericBounds,
+    pub Vec<P<AssocItem>>,
+);
```
```suggestion
);

```

bugadani commented:
> Can you add a static_assert_size! on Item, TraitItem, ImplItem and ForeignItem?

Sure, but I'm not sure if the structs have the same size on 32bit platforms  - wouldn't that be an issue, either now or later?

cjgillot commented:
> > Can you add a static_assert_size! on Item, TraitItem, ImplItem and ForeignItem?
> 
> Sure, but I'm not sure if the structs have the same size on 32bit platforms - wouldn't that be an issue, either now or later?

They wouldn't be the same. In `rustc_hir`, those asserts are only enabled on x86_64, and mostly serve as an indication to the reader.

bugadani commented:
@cjgillot I've added asserts (to the enums, because it's their sizes we want to keep low to avoid wasting memory for the "smaller" variants), and I've updated the PR with the size changes.

There are two more possibilities for the near future, based on this PR:

While gathering the sizes, I've noticed that some of the smaller variants in `ItemKind` are behind a pointer. I wonder if they are small enough to be inlined into the enum. I'll try that in a followup PR.

Also, #81400 showed that perf-wise there is not much difference between boxing a few variants vs boxing the whole enum. Since this PR aims at saving memory, I'll check if it's worth taking the current approach to the extremes and making `ItemKind` even smaller. This kind-of conflicts with the previous point, though.

cjgillot commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit 003fba3fdac2bcad6dbf067bc0e9eb0d1cdf7349 has been approved by `cjgillot`

<!-- @bors r=cjgillot 003fba3fdac2bcad6dbf067bc0e9eb0d1cdf7349 -->
<!-- homu: {"type":"Approved","sha":"003fba3fdac2bcad6dbf067bc0e9eb0d1cdf7349","approver":"cjgillot"} -->

bors commented:
:hourglass: Testing commit 003fba3fdac2bcad6dbf067bc0e9eb0d1cdf7349 with merge 3182375e064b8fa90437aee1465bccafd8187d89...
<!-- homu: {"type":"BuildStarted","head_sha":"003fba3fdac2bcad6dbf067bc0e9eb0d1cdf7349","merge_sha":"3182375e064b8fa90437aee1465bccafd8187d89"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1817613813)
Approved by: cjgillot
Pushing 3182375e064b8fa90437aee1465bccafd8187d89 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1817613813"},"merge_sha":"3182375e064b8fa90437aee1465bccafd8187d89"} -->
